### PR TITLE
BOT: Arcee and support

### PR DESCRIPTION
**Pull request by Northmoc**

### CODE REVIEW

**Comment by Northmoc** [comment]
wait until #1679 is merged

File: forge-gui/res/cardsfolder/upcoming/arcee_sharpshooter_arcee_acrobatic_coupe.txt
Comment (tool4ever) [inline forge-gui/res/cardsfolder/upcoming/arcee_sharpshooter_arcee_acrobatic_coupe.txt:23]: Uhh, from the wording I'd say you need to filter the amount too, e.g. a _Decimate_ shouldn't always count as 4

I don't think you'll need new AbilityKey anyway, something with `TriggerObjectsSpellAbilityTargetingCards$Valid` should work...

Comment (tool4ever) [reply forge-gui/res/cardsfolder/upcoming/arcee_sharpshooter_arcee_acrobatic_coupe.txt:23]: I found more problems:

- `SpellAbilityTargetingCards` only provides the targets of the root, that means you might miss some - currently only _Dack Fayden_ emblem is affected because on some others you used `RememberValidCards$` instead
but it would probably be better to remove that and use the fixed triggeringObject instead (`getAllTargetChoices`)
- The only other place where it gets used is this workaround: de8ff5a0e147cd8d2e27e68379a6614e0490ae78
I tested the case from that but it didn't seem needed, so we should probably remove it so it doesn't mess anything up
@Hanmac can you think of a scenario where we might lose the targets like that?
- `SpellAbilityTarget` for _Zevlor_ could also be broken if the targeting happens on a Sub and should be adressed similarly

Comment (Hanmac) [reply forge-gui/res/cardsfolder/upcoming/arcee_sharpshooter_arcee_acrobatic_coupe.txt:23]: first where is SpellAbilityTargetingCards really used? github can't show me. and should that be fixed to all targets instead?

The workaround looks ugly, is that really needed?

Comment (tool4ever) [reply forge-gui/res/cardsfolder/upcoming/arcee_sharpshooter_arcee_acrobatic_coupe.txt:23]: Only _Dack Fayden_ + `AbilityUtils` use it

but cards like _Orvar, the All-Form_ could also be refactored to use it instead of `RememberValidCards` imo

Comment (Hanmac) [reply forge-gui/res/cardsfolder/upcoming/arcee_sharpshooter_arcee_acrobatic_coupe.txt:23]: yeah from what i can see, they should use all targets ...
BUT probably should put it into a CardCollection to remove duplicates

File: forge-game/src/main/java/forge/game/card/Card.java
Comment (tool4ever) [inline forge-game/src/main/java/forge/game/card/Card.java:619]: part of these changes seem old/wrong again?

File: forge-game/src/main/java/forge/game/ability/AbilityKey.java
Comment (tool4ever) [inline forge-game/src/main/java/forge/game/ability/AbilityKey.java:122]: no longer needed?

File: forge-gui/res/cardsfolder/v/vesuvan_duplimancy.txt
Comment (tool4ever) [inline forge-gui/res/cardsfolder/v/vesuvan_duplimancy.txt:5]: This should already work with `TriggeredSpellAbilityTargetingCards`, so we can get rid of `SpellAbilityTarget` soon

File: forge-game/src/main/java/forge/game/trigger/TriggerSpellAbilityCastOrCopy.java
Comment (tool4ever) [inline forge-game/src/main/java/forge/game/trigger/TriggerSpellAbilityCastOrCopy.java:310]: it needs to be a CardCollection for better interaction with AbilityUtils

File: forge-game/src/main/java/forge/game/trigger/TriggerSpellAbilityCastOrCopy.java
Comment (tool4ever) [inline forge-game/src/main/java/forge/game/trigger/TriggerSpellAbilityCastOrCopy.java:309]: Probably
`TargetChoices firstTgt = null;`

and in the loop
```
if (!tc.isEmpty()) {
    firstTgt = tc;
}
```


```
if (firstTgt != null) {
    sa.setTriggeringObject(AbilityKey.SpellAbilityTarget, firstTgt.get(0));
}
```

File: forge-game/src/main/java/forge/game/card/Card.java
Comment (Northmoc) [reply forge-game/src/main/java/forge/game/card/Card.java]: I had a weird rebase in there, should be fixed now

**Review by tool4ever — DISMISSED** [review]
👍

File: forge-game/src/main/java/forge/game/trigger/TriggerSpellAbilityCastOrCopy.java
Comment (Hanmac) [inline forge-game/src/main/java/forge/game/trigger/TriggerSpellAbilityCastOrCopy.java:310]: wouldn't a FCollection<GameObject> or GameEntity be better?

then we maybe could get rid of AbilityKey.SpellAbilityTarget

Comment (Northmoc) [reply forge-game/src/main/java/forge/game/trigger/TriggerSpellAbilityCastOrCopy.java:310]: Worth a try, let's see!

Comment (Northmoc) [reply forge-game/src/main/java/forge/game/trigger/TriggerSpellAbilityCastOrCopy.java]: Seems to be working!

File: forge-game/src/main/java/forge/game/trigger/TriggerSpellAbilityCastOrCopy.java
Comment (Hanmac) [inline forge-game/src/main/java/forge/game/trigger/TriggerSpellAbilityCastOrCopy.java:308]: This needs to be an FCollection to skip duplicates

Comment (Northmoc) [reply forge-game/src/main/java/forge/game/trigger/TriggerSpellAbilityCastOrCopy.java:308]: Seems I can't do an `FCollection<TargetChoices>` ?

Comment (tool4ever) [reply forge-game/src/main/java/forge/game/trigger/TriggerSpellAbilityCastOrCopy.java]: It should probably be on the GameEntity level list though, since `getAllTargetChoices()` basically returns a List of Lists

otherwise some duplicates might still slip through

File: forge-gui/res/cardsfolder/z/zevlor_elturel_exile.txt
Comment (tool4ever) [inline forge-gui/res/cardsfolder/z/zevlor_elturel_exile.txt:8]: This got broken now
Try patching `AbilityUtils.getDefinedPlayers` for "Triggered" like this (instead of checking for CardCollection):
```
 else if (c instanceof Iterable<?>) { // For merged permanent
                    o = ((Iterable<?>) c).iterator().next();
                    if (orCont && o instanceof Card) {
                        o = ((Card) o).getController();
                    } 
                }
```

Comment (Northmoc) [reply forge-gui/res/cardsfolder/z/zevlor_elturel_exile.txt:8]: good catch

Comment (Northmoc) [reply forge-gui/res/cardsfolder/z/zevlor_elturel_exile.txt:8]: with this tweak, Zevlor seems to be working again

Comment (Hanmac) [reply forge-gui/res/cardsfolder/z/zevlor_elturel_exile.txt:8]: use the addPlayer function for Help, that can already support Iterable

Comment (Northmoc) [reply forge-gui/res/cardsfolder/z/zevlor_elturel_exile.txt:8]: Zevlor case: If I target a creature with Shock, it will make a copy for the creature's controller as well

Comment (Hanmac) [reply forge-gui/res/cardsfolder/z/zevlor_elturel_exile.txt:8]: the problem there is the OR Controller

you probably need to split it there:
```java
addPlayer(Iterables.filter(c, Card.class), 'Controller', players);
addPlayer(Iterables.filter(c, Player.class), '', players);
```

but shouldn't players be a PlayerCollection? it shouldn't be able to add multiples

Comment (Northmoc) [reply forge-gui/res/cardsfolder/z/zevlor_elturel_exile.txt:8]: Hmm, nothing but errors as I try that and similar: "incompatible equality constraints" and cast issues

Comment (tool4ever) [reply forge-gui/res/cardsfolder/z/zevlor_elturel_exile.txt:8]:
```
                    if (orCont) {
                        addPlayer(ImmutableList.copyOf(Iterables.filter((Iterable<Object>)c, Player.class)), "", players);
                    }
                    addPlayer(ImmutableList.copyOf(Iterables.filter((Iterable<Object>)c, Card.class)), "Controller", players);
```

The cast should be ok if that works

Comment (Northmoc) [reply forge-gui/res/cardsfolder/z/zevlor_elturel_exile.txt:8]: That seems to work!! Thank you!

**Review by tool4ever — APPROVED** [review]

**Pull request merged**
